modules/git/url/url.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package url
+
+import (
+	"fmt"
+	stdurl "net/url"
+	"strings"
+)
+
+// ErrWrongURLFormat represents an error with wrong url format
+type ErrWrongURLFormat struct {
+	URL string
+}
+
+func (err ErrWrongURLFormat) Error() string {
+	return fmt.Sprintf("git URL %s format is wrong", err.URL)
+}
+
+// GitURL represents a git URL
+type GitURL struct {
+	*stdurl.URL
+	extraMark int // 0 no extra 1 scp 2 file path with no prefix
+}
+
+// String returns the URL's string
+func (u *GitURL) String() string {
+	switch u.extraMark {
+	case 0:
+		return u.URL.String()
+	case 1:
+		return fmt.Sprintf("%s@%s:%s", u.User.Username(), u.Host, u.Path)
+	case 2:
+		return u.Path
+	default:
+		return ""
+	}
+}
+
+// Parse parse all kinds of git URL
+func Parse(remote string) (*GitURL, error) {
+	if strings.Contains(remote, "://") {
+		u, err := stdurl.Parse(remote)
+		if err != nil {
+			return nil, err
+		}
+		return &GitURL{URL: u}, nil
+	} else if strings.Contains(remote, "@") && strings.Contains(remote, ":") {
```

Comment:
The SCP target path parsing is found here: https://github.com/openssh/openssh-portable/blob/16ea8b85838dd7a4dbeba4e51ac4f43fd68b1e5b/misc.c#L767-L834

1. If there is a `'['` then we can split by the first `':'` after `']'` 
   - note this `':'` has to be immediately after the `]` otherwise the whole parse fails and it's a path.
2. Otherwise split by the first `':'`
3. We split by this `':'` into `<host_user> ':' <path>` 
   - if `<path>` is empty set to `"."`. If no `':'` set all to `<path>`
4. Split `<host_user>` by `'@'` into `<user> '@' <host-to-clean>`. 
   - if no `'@'` set all to `<host-to-clean>`
5.  `<host-to-clean>` - strip off `'['` and `']'` if present to form the `<host>`